### ISSUE TITLE
How to stack embedding and pass the gradients?

### ISSUE THREAD

**Issue by arita37** [TorchEnsemble-Community/Ensemble-Pytorch#108]
Have a 2 neural nets N1, N2,
want to stack their output embedding layer.

How to do this ?

**Comment by xuyxu** [comment]
Suppose the output of `N1` and `N2` on a sample `x` is `N1(x)` and `N2(x)` separately, what you want is to concatenate their output (i.e., `[N1(x), N2(x)]`), and pass it to downstream layers, right?

**Comment by arita37** [comment]
Exactly:
   Either concat, or   Mean 

We need to pass the gradient
and end to end training

**Comment by xuyxu** [comment]
Hi, if you are going to take the mean of outputs from all base estimators, the `fusion` ensemble is exactly what you want.

As to the concatenation, it is somehow weird since all base estimators in the ensemble are doing the same thing, making concatenating their outputs kind of useless. Is there any paper or technical report demonstrating the effectivenss of concatenating outputs of base estimators?

**Comment by arita37** [comment]
N1, N2,Nx  are different NN models.

We aggregate through concat their embedding output.

BigX = [ X1,….Xn]
and feed into another NN (ie merging).

This extensively used (ie Siamese Network…)

**Comment by arita37** [comment]
We are NOT dealing with output !!!

Output is kind of useless for End to End training…

We are dealing witht the last embedding.

**Comment by xuyxu** [comment]
Thanks for your kind explanation. Heterogeneous ensemble is not supported yet, since we have not come up with a succinct way on setting different optimizers for different base estimators 😢.

**Comment by xuyxu** [comment]
Kind of busy these days, will appreciate a PR very much ;-)
